### PR TITLE
Provide accessors for actions (rtnl_act).

### PR DESCRIPTION
**Pull request by shanmugh**
Add accessor functions to retrieve actions on a tc object
and provide additional function to navigate the collection
of actions.

### CODE REVIEW

File: lib/route/cls/basic.c [hunk 1]
```diff
@@ -233,6 +233,21 @@ int rtnl_basic_add_action(struct rtnl_cls *cls, struct rtnl_act *act)
...
+{
+	struct rtnl_basic *b;
+
+	if (!(b = rtnl_tc_data_peek(TC_CAST(cls))))
+		return NULL;
+
+	if (!(b->b_mask & BASIC_ATTR_ACTION))
+		return NULL;
+
+	struct rtnl_act* act = b->b_act;
```
Comment (congwang) [inline lib/route/cls/basic.c]: In C, it is preferred to to move all declarations to the beginning of the function.

File: lib/route/cls/basic.c [hunk 2]
```diff
@@ -233,6 +233,21 @@ int rtnl_basic_add_action(struct rtnl_cls *cls, struct rtnl_act *act)
...
+    rtnl_act_get(act);
```
Comment (congwang) [inline lib/route/cls/basic.c]: Fix indention

File: lib/route/cls/u32.c [hunk 3]
```diff
@@ -560,6 +560,21 @@ int rtnl_u32_add_action(struct rtnl_cls *cls, struct rtnl_act *act)
...
+    struct rtnl_u32 *u;
+
+    if (!(u = rtnl_tc_data_peek(TC_CAST(cls))))
+        return NULL;
+
+    if (!(u->cu_mask & U32_ATTR_ACTION))
+        return NULL;
+
+    struct rtnl_act* act = u->cu_act;
+    rtnl_act_get(act);
```
Comment (congwang) [inline lib/route/cls/u32.c]: This grabs a refcount, so you need to provide an API to release that refcount too?

File: libnl-route-3.sym [hunk 4]
```diff
@@ -822,6 +824,7 @@ global:
 	rtnl_u32_add_key_uint16;
 	rtnl_u32_add_key_uint32;
 	rtnl_u32_add_key_uint8;
+	rtnl_u32_get_action;
```
Comment (thom311) [inline libnl-route-3.sym]: let's insert new lines sorted alphabetically.

File: lib/route/cls/u32.c (hunk 3 above)
Comment (thom311) [inline lib/route/cls/u32.c]: That API exists as `rtnl_act_put()`. But I don't think that `get_action()` should increase the ref-count at all. Just return the reference without incrementing the ref-count.

Comment (shanmugh) [inline lib/route/cls/u32.c]: We already have `rtnl_put_act`.

File: lib/route/cls/basic.c (hunk 1 above)
Comment (shanmugh) [inline lib/route/cls/basic.c]: Fixed.

File: lib/route/cls/basic.c (hunk 2 above)
Comment (shanmugh) [inline lib/route/cls/basic.c]: Done.

File: libnl-route-3.sym (hunk 4 above)
Comment (shanmugh) [inline libnl-route-3.sym]: Fixed.

File: lib/route/cls/u32.c (hunk 3 above)
Comment (congwang) [inline lib/route/cls/u32.c]: @thom311 I mean if refcount is really needed rtnl_u32_put_action() is needed too, not that I miss rtnl_put_act()... But yeah, it looks like refcount is not necessary here.

**Comment by shanmugh** [comment]
The new API is needed to fix an issue in Mesos where we need to query the actions on a filter. See https://issues.apache.org/jira/browse/MESOS-7266.

**Comment by thom311** [comment]
Sorry for taking so long. I was rather busy these weeks.

I merged it now to master, but it missed 3.3.0 (which my fault). I didn't want to postpone 3.3.0 any longer and I also disliked the idea of adding new API after 3.3.0-rc1.

Let's see to do a 3.4.0 release earlier then planned.

**Pull request closed without merging**